### ISSUE TITLE
[tune] separate fixed configuration from searchable parameters

### ISSUE THREAD

**Issue by semin-park** [ray-project/ray#7215]
### Describe your feature request

I've wrapped an existing pytorch based project with ray tune PBT. It works great 👍thank you :)

But one thing I wish I had is, being able to input fixed configuration somewhere else other than in the configuration that is used for searching hyperparameters.

The things is, I have many options (e.g., dataset name and paths) that are completely irrelevant to the algorithm but are necessary. As far as I know, I have to include them in my hyperparameter space as list of length 1 in order to make them available to use in `_setup()`. But this looks messy and appears constantly when some hyperparameter perturbation occurs.

So it would be great if I could list my basic (fixed) configuration somewhere else, and use the hyperparameter space (dict) purely for tunable hyperparameters.

**Comment by richardliaw** [comment]
@semin-park, can you post an example of your ideal code?

**Comment by semin-park** [comment]
Sure :)

Currently my PBT perturbation log looks like this:

>>> 2020-02-20 03:29:24,873 INFO pbt.py:79 -- [explore] perturbed config from {'--dataset': 'NaverHackathonSpeech', '--data-path': PosixPath('/opt/data/naver-hackathon-speech/train'), '--split-prefix': PosixPath('/opt/data/naver-hackathon-speech/train'), '--train-path': PosixPath('train.csv'), '--val-path': PosixPath('val.csv'), '--cache-dir': PosixPath('/violin/.cache'), '--workers': 8, '--cpu': None, '--disable-amp': True, '--amp-level': 'O1', '--random-seed': 1, '--non-deterministic': None, '--subset-ratio': 0, '--skip-train': None, '--debug': None, '--epochs': 50, '--batch-size': 30, '--accumulate-grad-batches': 1, '--lr': 0.0014307782474712702, '--lr-scheduler': 'constant', '--gradient-clip-val': 5, '--weight-decay': True, '--weight-decay-value': 1e-05, '--dropout': 0.2, '--normalize-by-batch-size': None, '--tokenizer': 'char', '--vocabulary-path': PosixPath('/violin/asr/data/hackathon.labels'), '--audio-sample-rate': 16000, '--window-ms': 30, '--hop-ms': 10, '--remove-silence': True, '--feature': 'log-mel', '--n-fft': 512, '--n-mels': 80, '--n-mfcc': 80, '--delta': True, '--delta-delta': True, '--cmvn': None, '--augment': None, '--augment-resize-prob': None, '--augment-spec-prob': None, '--augment-spec-num': None, '--subsampler': 'vgg', '--encoder': 'rnn', '--decoder': 'rnn', '--rectangle-cnn-type': 'base', '--densenet-blocks': [6, 6, 6], '--densenet-growth-rate': 12, '--densenet-first-pooling': [2, 1], '--rnn-cell': 'gru', '--rnn-hidden-size': 512, '--hidden-size': 256, '--attention': 'global', '--transformer-d-model': 512, '--transformer-d-ff': 2048, '--transformer-n-heads': 8, '--num-layers': 3, '--layer-size': 3, '--num-encoder-layers': 3, '--num-decoder-layers': 1, '--max-length': 80, '--joint-ctc': 0.2, '--label-smoothing': None, '--decoding-sampling-rate': 0.5, '--search': 'greedy', '--beam-size': 1, '--beam-alpha': 0.2, '--beam-beta': 0.02, '--post-processing': None, 'asr-tune': True, '--perturb-freq': 2, '--population': 10, '--help': False, 'metric': 'val_cer'} -> {'--dataset': 'NaverHackathonSpeech', '--data-path': PosixPath('/opt/data/naver-hackathon-speech/train'), '--split-prefix': PosixPath('/opt/data/naver-hackathon-speech/train'), '--train-path': PosixPath('train.csv'), '--val-path': PosixPath('val.csv'), '--cache-dir': PosixPath('/violin/.cache'), '--workers': 8, '--cpu': None, '--disable-amp': True, '--amp-level': 'O1', '--random-seed': 1, '--non-deterministic': None, '--subset-ratio': 0, '--skip-train': None, '--debug': None, '--epochs': 50, '--batch-size': 30, '--accumulate-grad-batches': 1, '--lr': 0.0017169338969655241, '--lr-scheduler': 'constant', '--gradient-clip-val': 5, '--weight-decay': True, '--weight-decay-value': 1e-05, '--dropout': 0.2, '--normalize-by-batch-size': None, '--tokenizer': 'char', '--vocabulary-path': PosixPath('/violin/asr/data/hackathon.labels'), '--audio-sample-rate': 16000, '--window-ms': 30, '--hop-ms': 10, '--remove-silence': True, '--feature': 'log-mel', '--n-fft': 512, '--n-mels': 80, '--n-mfcc': 80, '--delta': True, '--delta-delta': True, '--cmvn': None, '--augment': None, '--augment-resize-prob': None, '--augment-spec-prob': None, '--augment-spec-num': None, '--subsampler': 'vgg', '--encoder': 'rnn', '--decoder': 'rnn', '--rectangle-cnn-type': 'base', '--densenet-blocks': [6, 6, 6], '--densenet-growth-rate': 12, '--densenet-first-pooling': [2, 1], '--rnn-cell': 'gru', '--rnn-hidden-size': 512, '--hidden-size': 256, '--attention': 'global', '--transformer-d-model': 512, '--transformer-d-ff': 2048, '--transformer-n-heads': 8, '--num-layers': 3, '--layer-size': 3, '--num-encoder-layers': 3, '--num-decoder-layers': 1, '--max-length': 80, '--joint-ctc': 0.2, '--label-smoothing': None, '--decoding-sampling-rate': 0.5, '--search': 'greedy', '--beam-size': 1, '--beam-alpha': 0.2, '--beam-beta': 0.02, '--post-processing': None, 'asr-tune': True, '--perturb-freq': 2, '--population': 10, '--help': False, 'metric': 'val_cer'}

It looks like I'm tuning many different hyperparameters, but in reality, all I'm changing is the learning rate. However, I have to put in all the other configuration in `tune.run(config=...)` because I need them all in order to create my model and trainer in `_setup`, and there is no other way of specifying those fixed arguments except for squishing them in `config`.

I can't put that information as an `__init__` argument either because it's not me who is constructing my `Trainable`. I tried declaring a global variable which holds my configuration before doing `ray.init()`, but for some reason it also didn't work.

So what I would like to have is another argument in `tune.run()` (possibly named `fixed_parameters` or something like that), where I can input all the miscellaneous arguments that are used in `_setup` but aren't related to the search space.

**Comment by richardliaw** [comment]
Are you using argparse? One option is to simply keep everything in the Namespace object and update it on the worker:

```
args = ArgumentParser().parse_args()
tune.run(Trainable, config={"base_args": args, "lr": ...})
```

BTW, this should also work:

```
base_config = {}

class Trainer(tune.Trainable):
    def _setup(self, config):
        self.base_config = base_config
        self.base_config.update(config)
    ...
```

**Comment by crypdick** [comment]
@richardliaw would it be difficult to add support for config to (optionally) be an argparse.Namespace? Otherwise, packing args into the config dict is awkward, e.g. `self.config["args"].max_epochs`. It seems like it would be as simple as adding `if isinstance(config, Namespace): config = vars(config).`

I wrote my codebase for args, so right now I do `args = config_to_args(config)` in each of my functions. 

```
def config_to_args(config):
  """
  ray tournament forces us to pass args inside dict
  let's make args first-class and unpack the dict into it
  """
  if isinstance(config, argparse.Namespace):
    args = config
  elif isinstance(config, dict):
    args = config.pop('args')
    for k, v in config.items():
      if hasattr(args, k):
        raise AttributeError(f"""key {k} was set in both
          args and Ray Tune config""")
      args.k = v
  else:
    raise TypeError("expecting type dict or argparser.Namespace, got {}")
  return args
```

**Comment by richardliaw** [comment]
yeah.. this should be easy to do.

**Comment by amogkam** [comment]
Hi @semin-park we changed this in Ray 1.0 so PBT only logs the hyperparameters that get mutated instead of the entire config. Give it a try!
